### ISSUE TITLE
Microsoft Teams Focused Even If No Window Is On Screen

### ISSUE THREAD

**Issue by nuynait** [koekeishiya/yabai#1172]
I'm having an issue that the Microsoft Teams app is taking focus even if no window is currently on the screen. 

# Steps

1. Have any window open. (But microsoft teams window all closed and app is opened for notification purpose.)
2. Focus on monitor-2
3. Focus back onto monitor-1

## Expected:
The window on monitor-1 gets focused.

## Actual
No window on monitor-1 gets focused. On the menu bar, I can see the current app has been switched to Microsoft Teams.

**Comment by nuynait** [comment]
My current temporary solution is I have a keybinding setting:
```
alt - tab : yabai -m window --focus "$(yabai -m query --windows --space | jq -re "[sort_by(.id, .frame) | reverse | .[] | select(.role == \"AXWindow\" and .subrole == \"AXStandardWindow\") | .id] | nth(index($(yabai -m query --windows --window | jq -re ".id")) - 1)")"
```

So everytime I focused to my main-monitor, I will have to press alt-tab to cycle the focus for once, then I can use opt-h/j/k/l to focus on the tile again. 

Are they better ways of doing this?

**Comment by Liquidmantis** [comment]
You're getting thwarted by the same "Microsoft Teams Notification" hidden window that causes me issues.  Inside your select filter, try adding `and .title != "Microsoft Teams Notification"`.

**Comment by typkrft** [comment]
> I bet you're getting thwarted by the same "Microsoft Teams Notification" hidden window that causes me issues. Inside your select filter, try adding `and .title != "Microsoft Teams Notification"`.

Recently Teams added native notifications. Not sure if that's whats going on here, but it might help.

**Comment by nuynait** [comment]
Yes. Changing the settings to native notification fixed the issue. I don't know that it's caused by the notification window and also not realize that it provides a native notification option here. Thanks so much for the help.

Since my issue has been solved, I will close this issue now.

**Comment by nuynait** [comment]
The problem comes back even after I change to the native notification. I think it is the meeting started notification that's doing the trick. Even if set mac native notification, it's still using its own notification window to display meeting notifications.

**Comment by typkrft** [comment]
You can run `sleep 2; yabai -m query --windows --window` on the popup window that seems to be the culprit. If you can differentiate that window from others you can try to make a rule not to manage it. Alternatively you can run teams from the browser, but I believe some stuff isn't browser compatible.

**Comment by nuynait** [comment]
@typkrft This is what I see when I run that command.

```
{
	"id":101025,
	"pid":50077,
	"app":"Microsoft Teams",
	"title":"Microsoft Teams Notification",
	"frame":{
		"x":2892.0000,
		"y":25.0000,
		"w":0.0000,
		"h":0.0000
	},
	"level":3,
	"role":"AXWindow",
	"subrole":"AXStandardWindow",
	"movable":1,
	"resizable":0,
	"display":1,
	"space":10,
	"visible":1,
	"focused":1,
	"split":"none",
	"floating":1,
	"sticky":1,
	"minimized":0,
	"topmost":1,
	"opacity":1.0000,
	"shadow":1,
	"border":1,
	"stack-index":0,
	"zoom-parent":0,
	"zoom-fullscreen":0,
	"native-fullscreen":0
}
```

Do you know what config I should have in `.yabairc` to disable management for this window? I tried `yabai -m rule --add app="^Microsoft Teams Notification$" manage=off` but it doesn't work.

**Comment by typkrft** [comment]
> Do you know what config I should have in `.yabairc` to disable management for this window? I tried `yabai -m rule --add app="^Microsoft Teams Notification$" manage=off` but it doesn't work.

You have `app` where it should say `title`.  You could add the `app` to, I saw an issue recently, though I can't remember the context in which someone said another rule wouldn't work without the `app`.

**Comment by shirdekel** [comment]
I also have this issue. Changing notification type and adding `yabai -m rule --add app="^Microsoft Teams$" title="^Microsoft Teams Notification$" manage=off` doesn't help.

**Comment by nyyManni** [comment]
For those willing to self-compile, I worked around this issue by treating the MS Notification as an unknown-type window:
https://github.com/nyyManni/yabai/commit/f979b7381de87a7a54f0e082ddb08578286ba947

I still seem to sometimes lose the focus to the notification when I close a window from the space, and none of the remaining windows gets focused.

**Comment by colossatr0n** [comment]
I've solved this issue by writing a script that runs when a Teams window is selected. It differentiates between the invisible notification window and an actual notification window. It also works across multiple displays. See  https://github.com/koekeishiya/yabai/issues/74#issuecomment-767904356.

As others have pointed out, changing to native Mac OS notifications only works for a little while. I've been using the script from the link above for about a year now. I had to update it today because yabai updated the names of some of its json fields.

**Comment by toastedonion** [comment]
FWIW, since I think this issue is more likely to be found, then the other one, I will link my comment here as well: https://github.com/koekeishiya/yabai/issues/74#issuecomment-1543681804

In principle, when changing the space and the `Microsoft Teams Notification` window gets focused it will simply look for other windows in that space and focus the first one it finds. You can read more details in the comment.

I haven't tested multiple displays, but I don't see a reason why it shouldn't work.

**Comment by fhenriquez** [comment]
Changing MS Teams notification style from "Teams built-in" to "Mac" fixes this issue.
